### PR TITLE
expose qr_thin_{Q,R}

### LINKED ISSUE #2560: add `qr_thin_{Q,R}` to language
#### Summary:

Stan Math now has the thin QR decomposition, but those functions have not been exposed to the Stan Language.

#### Description:

* [ ] add functions:

```
matrix qr_thin_Q(matrix);
matrix qr_thin_R(matrix);
```

* [ ] add tests (as indicated below) in the form of good model instantiations

* [ ] add doc to the function-signatures

#### Reproducible Steps:

`git grep -F "qr_thin" develop || echo "nothing found"`

#### Current Output:

`nothing found`

#### Expected Output:

```
src/docs/stan-reference/examples.tex:  matrix[K, K] R = qr_thin_R(x);
src/docs/stan-reference/examples.tex:  matrix[N, K] Q_ast = qr_thin_Q(x) * s;
src/stan/lang/function_signatures.h:add("qr_thin_Q", expr_type(matrix_type()), expr_type(matrix_type()));
src/stan/lang/function_signatures.h:add("qr_thin_R", expr_type(matrix_type()), expr_type(matrix_type()));
src/test/test-models/good/function-signatures/math/matrix/qr_thin_Q.stan:  transformed_data_matrix <- qr_thin_Q(d_matrix);
src/test/test-models/good/function-signatures/math/matrix/qr_thin_Q.stan:  transformed_param_matrix <- qr_thin_Q(d_matrix);
src/test/test-models/good/function-signatures/math/matrix/qr_thin_Q.stan:  transformed_param_matrix <- qr_thin_Q(p_matrix);
src/test/test-models/good/function-signatures/math/matrix/qr_thin_R.stan:  transformed_data_matrix <- qr_thin_R(d_matrix);
src/test/test-models/good/function-signatures/math/matrix/qr_thin_R.stan:  transformed_param_matrix <- qr_thin_R(d_matrix);
src/test/test-models/good/function-signatures/math/matrix/qr_thin_R.stan:  transformed_param_matrix <- qr_thin_R(p_matrix);
src/test/unit/lang/parser/matrix_functions_test.cpp:TEST(lang_parser, qr_thin_Q_matrix_function_signatures) {
src/test/unit/lang/parser/matrix_functions_test.cpp:  test_parsable("function-signatures/math/matrix/qr_thin_Q");
src/test/unit/lang/parser/matrix_functions_test.cpp:TEST(lang_parser, qr_thin_R_matrix_function_signatures) {
src/test/unit/lang/parser/matrix_functions_test.cpp:  test_parsable("function-signatures/math/matrix/qr_thin_R");
```

#### Additional Information:

This may introduce conflicts with the way files were moved around in #2558.  The function signatures doc stays the same, so if this only touches that and not the examples or reference manual, Git should be able to track the movements.

#### Current Version:
v2.17.1

### PR DESCRIPTION
**Pull request by bgoodri**
#### Submission Checklist

- [X] Run unit tests: `./runTests.py src/test/unit`
- [X] Run cpplint: `make cpplint`
- [X] Declare copyright holder and open-source license: see below

#### Summary

Expose `qr_thin_Q()` and `qr_thin_R()` to the Stan language

#### Intended Effect

Allow users to use `qr_thin_Q()` and `qr_thin_R()`.

Fixes #2560 .

#### How to Verify

Run tests under `src/test/test-models/good/function-signatures/math/matrix/`

#### Side Effects

Uses less RAM than fat QR decomposition

#### Documentation

Yes

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Trustees of Columbia University

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by bbbales2** [comment]
I'll check this tomorrow.

File: src/docs/stan-reference/functions.tex [hunk 1]
```diff
@@ -3131,15 +3140,21 @@ \subsubsection{QR Decomposition}\label{QR-decomposition}
 %
 \end{description}
 %
+The thin QR decomposition is generally preferable and will consume much
```
Comment (bbbales2) [inline src/docs/stan-reference/functions.tex]: Isn't it always preferred? Your choice, but shouldn't we just discourage folks from ever regular QRing in Stan, and then use the thin QR for the tupley QR whenever that happens? Is there any downside?

File: src/docs/stan-reference/functions.tex [hunk 2]
```diff
@@ -3119,6 +3119,15 @@ \subsubsection{QR Decomposition}\label{QR-decomposition}
 
 \begin{description}
 %
+\fitem{matrix}{qr\_thin\_Q}{matrix \farg{A}}{
+The orthogonal matrix in the thin QR decomposition of \farg{A}, which
+implies that the resulting matrix is the same dimensions as as \farg{A}}
+%
+\fitem{matrix}{qr\_thin\_R}{matrix \farg{A}}{
+  The upper triangular matrix in the thin QR decomposition of \farg{A},
+  which implies that the resulting matrix has the same dimensions as
```
Comment (bbbales2) [inline src/docs/stan-reference/functions.tex]: "the resulting matrix is square with the same number of columns as A."

Just to emphasize it is square so both dimensions are the same. This is the only necessary change.

**Comment by bbbales2** [comment]
I changed the docs as well. If you approve, say so and I'll merge. Otherwise edit away.

**Comment by bbbales2** [comment]
@seantalts Did the mechanism to retry Jenkins builds change? Went to restart this (there are only doc changes here so the tests should def. pass) but didn't find the "Build with Parameters" button.

**Comment by seantalts** [comment]
Sort of - There's a new Blue Ocean UI that now has a circle arrow icon for retrying in the top right if you're logged in. You can always go back to the old UI with the arrow in the box icon. More details will always be in this thread: http://discourse.mc-stan.org/t/new-jenkins-jobs-tutorial/2383

**Comment by bbbales2** [comment]
Yeah I wasn't logged in... My bad. Got it going again!

**Pull request merged**